tests/unit/s2n_random_test.c

```diff
@@ -184,169 +111,338 @@ int main(int argc, char **argv)
-            }
-        }
-        /* A common mistake in array filling leaves the last bytes zero
-         * depending on the length.
-         */
-        int remainder = i % 8;
-        int non_zero_found = 0;
-        for (int t = i - remainder; t < i; t++) {
-            non_zero_found |= data[t];
-        }
-        if (!non_zero_found) {
-            trailing_zeros[remainder]++;
-        }
+    if (random_func == GET_PUBLIC_RANDOM_DATA) {
+        EXPECT_OK(s2n_get_public_random_data(&thread_blob));
     }
-    for (int t = 1; t < 8; t++) {
-        EXPECT_TRUE(trailing_zeros[t] < 5120 / 16);
+    else if (random_func == GET_PRIVATE_RANDOM_DATA) {
+        EXPECT_OK(s2n_get_private_random_data(&thread_blob));
+    }
+    else {
+        EXPECT_SUCCESS(S2N_FAILURE);
     }
 
-    memset(trailing_zeros, 0, sizeof(trailing_zeros));
-    for (int i = 0; i < 5120; i++) {
-        blob.size = i;
-        EXPECT_OK(s2n_get_public_random_data(&blob));
+    EXPECT_OK(s2n_rand_cleanup_thread());
 
-        if (i >= 64) {
-            /* Set the run counts to 0 */
-            memset(bit_set_run, 0, 8);
+    return NULL;
+}
 
-            /* Apply 8 monobit tests to the data. Basically, we're
-             * looking for successive runs where a given bit is set.
-             * If a run exists with any particular bit 64 times in
-             * a row, then the data doesn't look randomly generated.
-             */
-            for (int j = 0; j < i; j++) {
-                for (int k = 0; k < 8; k++) {
-                    if (data[j] & bits[k]) {
-                        bit_set_run[k]++;
 
-                        if (j >= 64) {
-                            EXPECT_TRUE(bit_set_run[k] < 64);
-                        }
-                    } else {
-                        bit_set_run[k] = 0;
-                    }
-                }
-            }
-        }
-        /* A common mistake in array filling leaves the last bytes zero
-         * depending on the length.
-         */
-        int remainder = i % 8;
-        int non_zero_found = 0;
-        for (int t = i - remainder; t < i; t++) {
-            non_zero_found |= data[t];
-        }
-        if (!non_zero_found) {
-            trailing_zeros[remainder]++;
-        }
+static S2N_RESULT s2n_thread_test(S2N_RESULT (*s2n_get_random_data_cb)(struct s2n_blob *blob), uintptr_t thread_random_func)
+{
+    uint8_t data[RANDOM_GENERATE_DATA_SIZE];
+    struct s2n_blob blob = { .data = data };
+    pthread_t threads[MAX_NUMBER_OF_TEST_THREADS];
+
+    /* Create two threads and have them each grab RANDOM_GENERATE_DATA_SIZE
+     * bytes. The third parameter to pthread_create is packed. It containes two
+     * pieces of information: where to store the random data generated in the
+     * thread and which random function must be used to generate it.
+     */
+    EXPECT_EQUAL(pthread_create(&threads[0], NULL, s2n_thread_test_cb, (void *) (((uintptr_t) SLOT_NUM_0) | thread_random_func)), 0);
```

Comment:
The way you pass the args through seems pretty hacky. it's been a while since I worked with pthreads, but why can't you just pass a pointer to a struct with all the fields you want as the arg?